zebra/zapi_msg.c

```diff
@@ -1749,88 +1749,116 @@ static void zread_vrf_unregister(ZAPI_HANDLER_ARGS)
-	union g_addr gate;
-	ifindex_t ifindex;
-	mpls_label_t in_label, out_label;
-	uint8_t distance;
+	struct zapi_labels zl;
 
 	/* Get input stream.  */
 	s = msg;
+	if (zapi_labels_decode(s, &zl) < 0) {
+		if (IS_ZEBRA_DEBUG_RECV)
+			zlog_debug("%s: Unable to decode zapi_labels sent",
+				   __PRETTY_FUNCTION__);
+		return;
+	}
 
-	/* Get data. */
-	STREAM_GETC(s, type);
-	STREAM_GETL(s, prefix.family);
-	switch (prefix.family) {
-	case AF_INET:
-		STREAM_GET(&prefix.u.prefix4.s_addr, s, IPV4_MAX_BYTELEN);
-		STREAM_GETC(s, prefix.prefixlen);
-		if (prefix.prefixlen > IPV4_MAX_BITLEN) {
-			zlog_debug(
-				"%s: Specified prefix length %d is greater than a v4 address can support",
-				__PRETTY_FUNCTION__, prefix.prefixlen);
-			return;
-		}
-		STREAM_GET(&gate.ipv4.s_addr, s, IPV4_MAX_BYTELEN);
-		break;
-	case AF_INET6:
-		STREAM_GET(&prefix.u.prefix6, s, 16);
-		STREAM_GETC(s, prefix.prefixlen);
-		if (prefix.prefixlen > IPV6_MAX_BITLEN) {
-			zlog_debug(
-				"%s: Specified prefix length %d is greater than a v6 address can support",
-				__PRETTY_FUNCTION__, prefix.prefixlen);
-			return;
-		}
-		STREAM_GET(&gate.ipv6, s, 16);
-		break;
-	default:
-		zlog_debug("%s: Specified AF %d is not supported for this call",
-			   __PRETTY_FUNCTION__, prefix.family);
+	if (!mpls_enabled)
 		return;
+
+	for (int i = 0; i < zl.nexthop_num; i++) {
+		struct zapi_nexthop_label *znh;
+
+		znh = &zl.nexthops[i];
+		mpls_lsp_install(zvrf, zl.type, zl.local_label, znh->label,
+				 znh->type, &znh->address, znh->ifindex);
+
+		if (CHECK_FLAG(zl.message, ZAPI_LABELS_FTN))
+			mpls_ftn_update(1, zvrf, zl.type, &zl.route.prefix,
+					znh->type, &znh->address, znh->ifindex,
+					zl.route.type, zl.route.instance,
+					znh->label);
 	}
-	STREAM_GETL(s, ifindex);
-	STREAM_GETC(s, distance);
-	STREAM_GETL(s, in_label);
-	STREAM_GETL(s, out_label);
+}
 
-	switch (prefix.family) {
-	case AF_INET:
-		if (ifindex)
-			gtype = NEXTHOP_TYPE_IPV4_IFINDEX;
-		else
-			gtype = NEXTHOP_TYPE_IPV4;
-		break;
-	case AF_INET6:
-		if (ifindex)
-			gtype = NEXTHOP_TYPE_IPV6_IFINDEX;
-		else
-			gtype = NEXTHOP_TYPE_IPV6;
-		break;
-	default:
+static void zread_mpls_labels_replace(ZAPI_HANDLER_ARGS)
+{
+	struct stream *s;
+	struct zapi_labels zl;
+
+	/* Get input stream.  */
+	s = msg;
+	if (zapi_labels_decode(s, &zl) < 0) {
+		if (IS_ZEBRA_DEBUG_RECV)
+			zlog_debug("%s: Unable to decode zapi_labels sent",
+				   __PRETTY_FUNCTION__);
 		return;
 	}
 
 	if (!mpls_enabled)
 		return;
 
-	if (hdr->command == ZEBRA_MPLS_LABELS_ADD) {
-		mpls_lsp_install(zvrf, type, in_label, out_label, gtype, &gate,
-				 ifindex);
-		mpls_ftn_update(1, zvrf, type, &prefix, gtype, &gate, ifindex,
-				distance, out_label);
-	} else if (hdr->command == ZEBRA_MPLS_LABELS_DELETE) {
-		mpls_lsp_uninstall(zvrf, type, in_label, gtype, &gate, ifindex);
-		mpls_ftn_update(0, zvrf, type, &prefix, gtype, &gate, ifindex,
-				distance, out_label);
+	mpls_lsp_uninstall_all_vrf(zvrf, zl.type, zl.local_label);
```

Comment:
As this function intent to work in `Make Before Break` mode, how can you guarantee that deletion inserted in the command queue will be overwrite by subsequent new configuration ? Is there a risk that deletion will be executed before addition ?

Reply:
When `mpls_lsp_uninstall_all_vrf()` is called here, that function doesn't actually remove/uninstall anything. It only sets the `NHLFE_FLAG_DELETED` flag on all NHLFEs associated to the LSP, then queues this LSP for further processing on a separate event callback.

Later, `zread_mpls_labels_replace()` calls `mpls_lsp_install()`, which does the inverse thing. It creates new NHLFEs or unsets the `NHLFE_FLAG_DELETED` flag of existing NHLFEs.

When the LSP is processed later on a separate event callback, zebra will check which NHLFEs are still active and reinstall the LSP based on that (using route replace semantics if available by the underlying operating system). For more information, you can take a look at the `lsp_process()` and `lsp_select_best_nhlfe()` functions.

To summarize, there's no risk that the LSP can be uninstalled from the forwarding plane before being updated with the new NHLFEs.